thorlcr/activities/indexread/thindexreadslave.cpp

```diff
@@ -89,36 +98,202 @@ class CIndexReadSlaveBase : public CSlaveActivity
         {
             keyManager = NULL;
         }
     } callback;
 
-    virtual bool keyed() { return false; }
-    virtual void setManager(IKeyManager *manager)
+    // return a ITranslator based on published format in part and expected/format
+    ITranslator *getTranslators(IPartDescriptor &partDesc)
     {
-        clearManager();
-        currentManager = manager;
-        callback.setManager(manager);
-        resetLastStats();
-        helper->createSegmentMonitors(currentManager);
-        currentManager->finishSegmentMonitors();
-        currentManager->reset();
-        keyMergerInUse = (manager == keyMergerManager);
+        unsigned expectedFormatCrc = helper->getFormatCrc();
+        IOutputMetaData *projectedFormat = helper->queryProjectedDiskRecordSize();
+        IPropertyTree const &props = partDesc.queryOwner().queryProperties();
+        Owned<IOutputMetaData> publishedFormat = getDaliLayoutInfo(props);
+        unsigned publishedFormatCrc = (unsigned)props.getPropInt("@formatCrc", 0);
+        RecordTranslationMode translationMode = getTranslationMode(*this);
+        IOutputMetaData *expectedFormat = helper->queryDiskRecordSize();
+
+        Owned<ITranslator> ret = ::getTranslators("rowstream", expectedFormat, publishedFormat, projectedFormat, translationMode, expectedFormatCrc, false, publishedFormatCrc);
+        if (!ret)
+            return nullptr;
+        if (!ret->queryTranslator().canTranslate())
+            throw MakeStringException(0, "Untranslatable key layout mismatch reading index %s", logicalFilename.get());
+        if (ret->queryTranslator().keyedTranslated())
+            throw MakeStringException(0, "Untranslatable key layout mismatch reading index %s - keyed fields do not match", logicalFilename.get());
+        return ret.getClear();
+    }
+public:
+    IIndexLookup *getNextInput(IKeyManager *&keyManager, unsigned &partNum, bool useMerger)
+    {
+        keyManager = nullptr;
+        if (useMerger && keyMergerManager)
+        {
+            if (0 == partNum)
+                keyManager = keyMergerManager;
+        }
+        else if (keyManagers.isItem(partNum))
+            keyManager = &keyManagers.item(partNum);
+        else if (partNum >= partDescs.ordinality())
+            return nullptr;
+        else
+        {
+            if (localMerge) // NB: keyManagers[0] will be filled below
+            {
+                if (partNum > 0)
+                    return nullptr;
+            }
+            RecordTranslationMode translationMode = getTranslationMode(*this);
+            IOutputMetaData *projectedFormat = helper->queryProjectedDiskRecordSize();
+            unsigned expectedFormatCrc = helper->getFormatCrc();
+
+            unsigned p = partNum;
+            while (p<partDescs.ordinality()) // will process all parts if localMerge
+            {
+                IPartDescriptor &part = partDescs.item(p++);
+                for (unsigned copy=0; copy<part.numCopies(); copy++)
```

Comment:
I can't see how you can go round this loop more than once.  I think the tail of the loop either breaks or returns.

Reply:
You're right, I'll look at this further